File: binary-search/5LongestPalindromicSubstring.java
class Solution {
    public String longestPalindrome(String s) {
        if(s == null || s.length() < 1) return "";
        int l = 0, r = 0;
        for(int i = 0; i < s.length(); i++) {
            int len1 = expandAroundCenter(i, i, s);
            int len2 = expandAroundCenter(i, i + 1, s);
            int len = Math.max(len1, len2);
            if(len > r - l) {
                l = i - (len - 1) / 2;
                r = i + len / 2;
            }
        }
    }

    public int expandAroundCenter(int l, int r, String s) {
        while(l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l++;
            r--;
        }
        return r - l - 1;
    }
}

File: two-pointers/5LongestPalindromicSubstring.py
def longestPalindrome(s):

    def expandAroundCenter(s, l, r):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return r - l - 1

    if not s or len(s) == 0: return ""
    l, r = 0, 0
    for i in range(len(s)):
        l1 = expandAroundCenter(s, i, i)
        l2 = expandAroundCenter(s, i, i+1)
        m = max(l1, l2)
        if m > r - l:
            l = i - (m - 1) / 2
            r = i + m / 2
    return s[l:r]